### PR TITLE
[WB-6966] yolov5-network-usage-bug

### PR DESCRIPTION
**Pull request by vwrj**
https://wandb.atlassian.net/browse/WB-6966
https://github.com/ultralytics/yolov5/issues/5071

W&B is sending a large amount of network egress for yolov5 repo on its Docker image on GCP instance. 

When reproducing, I found a bug with our console streaming that also caused a large amount of network egress:

- Tqdm progress bars write to the STDERR stream. If a user prints to STDOUT during these progress bar updates, our console streaming code keeps appending new lines and sending this growing set of lines with *every* POST request.
- Effectively we end up sending the growing output.log file with every update. 

The below figure was generated for a 10-epoch MNIST run. The `train` function was called every epoch. Inside, it used a tqdm progress bar and printed the train loss every 10 minibatches to STDOUT. 

`output.log` file was only ~250 KB. 

cumulative traffic (mb) to _files endpoint (1)

Description

-----------

This PR makes significant changes to `CRDedupeFilePolicy`, more specifically the `process_chunks` function.

For each stream (stderr & stdout), it keeps track of the latest offset with `\r` and the latest normal line.  Any time it encounters a line with a carriage return `\r`, it maps it to the correct offset, thereby overwriting the line. Every progress bar line (except for the first one) has a `\r`. Finally at the end, it packages consecutive lines together and makes a POST request for each consecutive list of lines. 

It only sends the current chunk of lines it processes; therefore it will never send more data than it receives in the current chunk. 

Here's the same chart as above, with the new process_chunks function using the same MNIST train script as above. 

cumulative traffic (mb) to _files endpoint (2)

Testing
-------

Unittests for each of the convenience functions used in `CRDedupeFilePolicy`: `split_chunk` and `get_consecutive_offsets`. 
A functional test for `process_chunks` that instantiates the FilePolicy object and sends it a series of three sets of chunks, intended to simulate progress bars for both STDERR and STDOUT streams. 

Release Notes
-------------

------------- BEGIN RELEASE NOTES ------------------
Fix issue where console log streaming was causing excessive network traffic. 
------------- END RELEASE NOTES --------------------

### CODE REVIEW

File: wandb/sdk/lib/file_stream_utils.py [hunk 1]
```diff
@@ -50,13 +51,26 @@ def _num_lines_from_num_bytes(file, num_bytes):
             num_lines += 1
         return num_lines
 
-    files_stack = [
-        {"name": k, "offset": v["offset"], "content": v["content"]}
-        for k, v in files.items()
-    ]
+    files_stack = []
+    for k, v in files.items():
+        if type(v) is list:
```
Comment (dannygoldstein) [inline wandb/sdk/lib/file_stream_utils.py:56]:
```suggestion
        if isinstance(v, list):
```

File: wandb/sdk/internal/file_stream.py [hunk 2]
```diff
@@ -71,66 +71,114 @@ def process_chunks(self, chunks):
...
+        # after for loop: keys = [(2, 4), (8, 10), (20, 20)]
+        for _, group in itertools.groupby(enumerate(offsets), lambda x: x[0] - x[1]):
+            group = list(map(lambda x: x[1], group))
+            item = (group[0], group[-1]) if len(group) > 1 else (group[0], group[0])
+            intervals.append(item)
+        return intervals
+
+    def split_chunk(self, chunk):
+        """
+        chunk: object of type Chunk with two fields: filename (str) & data (str)
```
Comment (dannygoldstein) [inline wandb/sdk/internal/file_stream.py:128]: Looks like the definition of `Chunk` above has only `data`?

**Review by dannygoldstein — APPROVED** [review]
lgtm

**Review by vanpelt — APPROVED** [review]
Really clean fix and great tests.  Nice work!

**Pull request merged**